src/common/src/hash/table_distribution.rs

```diff
@@ -0,0 +1,283 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::mem::replace;
+use std::ops::Deref;
+use std::sync::{Arc, LazyLock};
+
+use itertools::Itertools;
+use risingwave_pb::plan_common::StorageTableDesc;
+use tracing::warn;
+
+use crate::array::{Array, DataChunk, PrimitiveArray};
+use crate::buffer::{Bitmap, BitmapBuilder};
+use crate::hash::VirtualNode;
+use crate::row::Row;
+use crate::util::iter_util::ZipEqFast;
+
+/// For tables without distribution (singleton), the `DEFAULT_VNODE` is encoded.
+pub const DEFAULT_VNODE: VirtualNode = VirtualNode::ZERO;
+
+#[derive(Debug, Clone)]
+enum ComputeVnode {
+    Singleton,
+    DistKeyIndices {
+        /// Indices of distribution key for computing vnode, based on the all columns of the table.
+        dist_key_in_pk_indices: Vec<usize>,
+    },
+    VnodeColumnIndex {
+        /// Indices of vnode columns.
+        vnode_col_idx_in_pk: usize,
+    },
+}
+
+#[derive(Debug, Clone)]
+/// Represents the distribution for a specific table instance.
+pub struct TableDistribution {
```

Comment:
Will we also use it to rewrite the streaming `StateTable`?

Reply:
The `StateTable` and `StorageTable` have already used it in some previous refactor PRs. It is moved from `risingwave_storage` crate to the common crate so that the frontend can use it.